net/dns/resolver/forwarder.go

```diff
@@ -49,6 +50,11 @@ const (
 	// arbitrary.
 	dohTransportTimeout = 30 * time.Second
 
+	// dohTransportTimeout is how much of a head start to give a DoH query
+	// that was upgraded from a well-known public DNS provider's IP before
+	// normal UDP mode is attempted as a fallback.
+	dohHeadStart = 1500 * time.Millisecond
```

Comment:
1.5s delays when DoH is sad is about as bad as DNS being down, from a UX POV. Any chance we can break this down into a first-connection-setup delay (where 1.5s is okay), vs. a request over an already connected channel (where we'd probably want to try fallbacks sooner) ?

Reply:
I'm open to other values here.

Deciding based on whether it's connected or not is pretty tricky, though. It's less work (but still annoying) to know whether there's definitely not a connection, but when there is a TCP connection open, it's not clear whether it'll work right away or time out and work later, etc.

I'd rather we just lower this for now. But to what?

Reply:
500ms?

Reply:
I'm not sure if we are able to access `net.Conn` underneath all the various abstractions, but could we use `SetReadDeadline()` + `SetWriteDeadline()` to cause the request to error faster in the case we are re-using an existing connection?

Reply:
it's an http.Client -> http.Transport -> http2.Transport -> h2ConnPool -> maybe more than 1 net.Conn!

we could use the various ClientTrace hooks (https://pkg.go.dev/net/http/httptrace#ClientTrace) to see some of the lifecycle, etc.

Reply:
changed to 500ms for now. still an eternity compared to how quickly I see Google/Cloudflare complete a cold DoH query anyway.